### PR TITLE
Implement DictionarySegments using std::span

### PR DESCRIPTION
**Pull request by tjjordan**
Note: The content of this PR still needs to undergo major refactorings.

This PR introduces `DictionarySegments` that make use of `std::span`.
If they are created "traditionally" they still store their data inside vectors, but all work they do is performed on `std::spans` spanning on the data of the vectors.
To create `DictionarySegments` on mapped data, they now own a static `DictionarySegment<T>::create()` method, expecting a map and an offset.
This method creates an DictionarySegment owningspans spanning over the mapped data. The segment's vectors stay uninitialized in this case.

### CODE REVIEW

**Comment by Bouncner** [comment]
Why `DictionarySegment<T>::create()` instead of another constructor?

File: src/lib/storage/dictionary_segment.cpp [hunk 1]
```diff
@@ -16,15 +16,78 @@ DictionarySegment<T>::DictionarySegment(const std::shared_ptr<const pmr_vector<T
...
+    const auto dictionary_size = map[segment_start_map_index];
+    const auto attribute_vector_size = map[segment_start_map_index + 1];
+    const auto encoding_type = map[segment_start_map_index + 2];
+    const auto type_size_as_index = sizeof(T) / 4;
+
+    auto* dictionary_map = reinterpret_cast<const T*>(map);
+    auto dictionary_span = std::span<const T>(&dictionary_map[segment_start_map_index + 3], dictionary_size);
+    auto dictionary_span_pointer = std::make_shared<std::span<const T>>(dictionary_span);
+
+    if (encoding_type == 1) {
```
Comment (Bouncner) [inline src/lib/storage/dictionary_segment.cpp:56]: Better instantiate a real `EncodingType` and use a switch.

File: src/lib/storage/vector_compression/bitpacking/bitpacking_vector.cpp [hunk 2]
```diff
@@ -11,6 +11,10 @@ const pmr_compact_vector& BitPackingVector::data() const {
   return _data;
 }
 
+void* BitPackingVector::data_pointer() const {
```
Comment (Bouncner) [inline src/lib/storage/vector_compression/bitpacking/bitpacking_vector.cpp:14]: I don't see a usage of that method. Is it a left-over?

File: src/lib/storage/create_iterable_from_segment.ipp [hunk 3]
```diff
@@ -50,12 +50,12 @@ template <typename T, bool EraseSegmentType>
 auto create_iterable_from_segment(const FixedStringDictionarySegment<T>& segment) {
 #ifdef HYRISE_ERASE_FIXEDSTRINGDICTIONARY
   PerformanceWarning("FixedStringDictionarySegmentIterable erased by compile-time setting");
-  return AnySegmentIterable<T>(DictionarySegmentIterable<T, FixedStringVector>(segment));
+  return AnySegmentIterable<T>(DictionarySegmentIterable<T, FixedStringVector, std::span<const int>>(segment));
```
Comment (Bouncner) [inline src/lib/storage/create_iterable_from_segment.ipp:53]: Once "we are done" with the dictionaries, the second template parameter can get removed?

**Review by Bouncner — COMMENTED** [review]
Just a few comments. Hard to see what are "hacks", left-overs, or what is going to stay.

But looks promising. In general, do you see any issues that  might stop us in the long run?

File: src/lib/storage/dictionary_segment.cpp (hunk 1 above)
Comment (SanJSp) [reply src/lib/storage/dictionary_segment.cpp:56]: Something similar already exists in `StorageManager.hpp`, might be interesting

File: src/lib/storage/dictionary_segment.cpp [hunk 4]
```diff
@@ -16,15 +16,78 @@ DictionarySegment<T>::DictionarySegment(const std::shared_ptr<const pmr_vector<T
...
+      auto* attribute_vector_map = reinterpret_cast<const uint8_t*>(map);
+      auto attribute_data_span = std::span<const uint8_t>(
+          &attribute_vector_map[(segment_start_map_index + 3 + dictionary_size * type_size_as_index) * 4],
+          attribute_vector_size);
```
Comment (SanJSp) [inline src/lib/storage/dictionary_segment.cpp:60]: magic numbers (+3; * 4)

File: src/lib/storage/dictionary_segment.cpp [hunk 5]
```diff
@@ -16,15 +16,78 @@ DictionarySegment<T>::DictionarySegment(const std::shared_ptr<const pmr_vector<T
...
+
+      return std::make_shared<DictionarySegment<T>>(dictionary_span_pointer, attribute_vector);
+    } else if (encoding_type == 4) {
+      Fail("BitPacking not implemented yet.");
+    } else {
+      Fail("Unsupported EncodingType.");
+    }
+
+  } else {
+    Fail("Das tun wir hier nicht.");
```
Comment (SanJSp) [inline src/lib/storage/dictionary_segment.cpp:87]: What kind of fail message is that?

File: src/lib/storage/storage_manager.cpp [hunk 6]
```diff
@@ -503,53 +505,20 @@ std::shared_ptr<Chunk> StorageManager::map_chunk_from_disk(const uint32_t chunk_
 
   // TODO: Remove unneccesary map on whole file
   const auto* map = reinterpret_cast<uint32_t*>(mmap(NULL, file_bytes, PROT_READ, MAP_PRIVATE, fd, off_t{0}));
+  _maps.emplace_back(map);
```
Comment (SanJSp) [inline src/lib/storage/storage_manager.cpp:508]: maybe we should store here a reference to what is stored in this map? Like f.e. the filename could work for now but in the future we should use a better DS here

File: src/lib/storage/dictionary_segment.cpp [hunk 7]
```diff
@@ -16,15 +16,78 @@ DictionarySegment<T>::DictionarySegment(const std::shared_ptr<const pmr_vector<T
...
+      auto attribute_data_span = std::span<const uint8_t>(
+          &attribute_vector_map[(segment_start_map_index + 3 + dictionary_size * type_size_as_index) * 4],
+          attribute_vector_size);
+      auto attribute_vector = std::make_shared<FixedWidthIntegerVector<uint8_t>>(attribute_data_span);
+
+      return std::make_shared<DictionarySegment<T>>(dictionary_span_pointer, attribute_vector);
+    } else if (encoding_type == 2) {
+      auto* attribute_vector_map = reinterpret_cast<const uint16_t*>(map);
+      auto attribute_data_span = std::span<const uint16_t>(
+          &attribute_vector_map[(segment_start_map_index + 3 + dictionary_size * type_size_as_index) * 2],
```
Comment (SanJSp) [inline src/lib/storage/dictionary_segment.cpp:67]: again magic numbers

**Review by SanJSp — CHANGES_REQUESTED** [review]
Good Work!!

Some minor comments from my side - maybe Theresa is the best to take another look at this, to check if this aligns with her expectations of the implementation

File: src/lib/storage/vector_compression/bitpacking/bitpacking_vector.cpp (hunk 2 above)
Comment (tjjordan) [reply src/lib/storage/vector_compression/bitpacking/bitpacking_vector.cpp:14]: I think it is supposed to be the `std::span` alternative to the `.data()` method of the `BitPackingVector`.
I'll remove it for now until it is needed.

File: src/lib/storage/create_iterable_from_segment.ipp (hunk 3 above)
Comment (tjjordan) [reply src/lib/storage/create_iterable_from_segment.ipp:53]: If we get rid of vectors inside the segments then yes.

File: src/lib/storage/dictionary_segment.cpp (hunk 4 above)
Comment (tjjordan) [reply src/lib/storage/dictionary_segment.cpp]: How to do this cleaner?

File: src/lib/storage/dictionary_segment.cpp (hunk 5 above)
Comment (tjjordan) [reply src/lib/storage/dictionary_segment.cpp:109]: A funny one.

File: src/lib/storage/storage_manager.cpp (hunk 6 above)
Comment (tjjordan) [reply src/lib/storage/storage_manager.cpp:508]: We are also not supposed to use raw_pointers.
But I think this map is not in the scope of this PR.

**Review by rrcomtech — DISMISSED** [review]
As was already pointed out by @tjjordan, this is a pr-based pr and the actual discussion should take place on `file_format_refactored`. Therefore, lets get this merged and continue discussing the interface on the other branch.

File: src/lib/storage/dictionary_segment.hpp [hunk 8]
```diff
@@ -22,9 +23,17 @@ class DictionarySegment : public BaseDictionarySegment {
   explicit DictionarySegment(const std::shared_ptr<const pmr_vector<T>>& dictionary,
                              const std::shared_ptr<const BaseCompressedVector>& attribute_vector);
 
+  explicit DictionarySegment(const std::shared_ptr<const std::span<const T>>& dictionary,
+                             const std::shared_ptr<const BaseCompressedVector>& attribute_vector);
+
+  explicit DictionarySegment(const uint32_t* map, const uint32_t segment_start_offset_bytes);
```
Comment (mweisgut) [inline src/lib/storage/dictionary_segment.hpp:29]: Same as above, `start_address` or `base_address` would be more descriptive IMO.

File: src/lib/storage/dictionary_segment/dictionary_segment_iterable.hpp [hunk 9]
```diff
@@ -28,12 +29,13 @@ class DictionarySegmentIterable : public PointAccessibleSegmentIterable<Dictiona
...
       using CompressedVectorIterator = decltype(vector.cbegin());
-      using DictionaryIteratorType = decltype(_dictionary->cbegin());
+      using DictionaryIteratorType = decltype(_dictionary_span->begin());
 
       auto begin = Iterator<CompressedVectorIterator, DictionaryIteratorType>{
-          _dictionary->cbegin(), _segment.null_value_id(), vector.cbegin(), ChunkOffset{0u}};
-      auto end = Iterator<CompressedVectorIterator, DictionaryIteratorType>{
-          _dictionary->cbegin(), _segment.null_value_id(), vector.cend(), static_cast<ChunkOffset>(_segment.size())};
+          _dictionary_span->begin(), _segment.null_value_id(), vector.cbegin(), ChunkOffset{0u}};
+      auto end = Iterator<CompressedVectorIterator, DictionaryIteratorType>{_dictionary_span->begin(),
```
Comment (mweisgut) [inline src/lib/storage/dictionary_segment/dictionary_segment_iterable.hpp:36]: Shouldn't we use `_dictionary_span->end()` here?

File: src/lib/storage/storage_manager.hpp [hunk 10]
```diff
@@ -104,6 +111,7 @@ class StorageManager : public Noncopyable {
   tbb::concurrent_unordered_map<std::string, std::shared_ptr<Table>> _tables{INITIAL_MAP_SIZE};
   tbb::concurrent_unordered_map<std::string, std::shared_ptr<LQPView>> _views{INITIAL_MAP_SIZE};
   tbb::concurrent_unordered_map<std::string, std::shared_ptr<PreparedPlan>> _prepared_plans{INITIAL_MAP_SIZE};
+  std::vector<const uint32_t*> _maps;
```
Comment (mweisgut) [inline src/lib/storage/storage_manager.hpp:114]: When we use `MemoryRegion`s or something similar, we will store a vector of `MemoryRegions` here. Nevertheless, the naming can be more descriptive here, indicating that we store start addresses of memory regions here.

File: src/lib/storage/storage_manager.hpp [hunk 11]
```diff
@@ -127,10 +135,11 @@ class StorageManager : public Noncopyable {
...
 
   // Segment Header
-    static const uint32_t _dictionary_size_bytes = 4;
-    static const uint32_t _element_count_bytes = 4;
-    static const uint32_t _compressed_vector_type_id_bytes = 4;
-    static const uint32_t _segment_header_bytes = _dictionary_size_bytes + _element_count_bytes + _compressed_vector_type_id_bytes;
+  static const uint32_t _dictionary_size_bytes = 4;
+  static const uint32_t _element_count_bytes = 4;
+  static const uint32_t _compressed_vector_type_id_bytes = 4;
+  static const uint32_t _segment_header_bytes =
```
Comment (mweisgut) [inline src/lib/storage/storage_manager.hpp:141]: Can these be `static constexpr`?

File: src/test/lib/storage/storage_manager_test.cpp [hunk 12]
```diff
@@ -292,4 +292,120 @@ TEST_F(StorageManagerTest, WriteMaxNumberOfChunksToFile) {
...
 }
 
+TEST_F(StorageManagerTest, WriteMaxNumberOfChunksToFileSmall) {
+  const auto file_name = "test_chunks_file.bin";
+  std::remove(file_name);
+  auto& sm = Hyrise::get().storage_manager;
+
+  const auto ROW_COUNT = uint32_t{100};  // can't be greater than INT32_MAX
+  const auto COLUMN_COUNT = uint32_t{23};
+  const auto CHUNK_COUNT = sm.get_max_chunk_count_per_file();
```
Comment (mweisgut) [inline src/test/lib/storage/storage_manager_test.cpp:302]: `constexpr` possible?

File: src/lib/storage/dictionary_segment/dictionary_segment_iterable.hpp [hunk 13]
```diff
@@ -10,16 +10,17 @@
 
 namespace hyrise {
 
-template <typename T, typename Dictionary>
-class DictionarySegmentIterable : public PointAccessibleSegmentIterable<DictionarySegmentIterable<T, Dictionary>> {
+template <typename T, typename Dictionary, typename DictionarySpan>
+class DictionarySegmentIterable
+    : public PointAccessibleSegmentIterable<DictionarySegmentIterable<T, Dictionary, DictionarySpan>> {
```
Comment (klauck) [inline src/lib/storage/dictionary_segment/dictionary_segment_iterable.hpp:15]: Will the `_dictionary` member be removed in the future or is the `_dictionary_span` lacking required features?

File: src/lib/storage/dictionary_segment.hpp [hunk 14]
```diff
@@ -86,7 +95,8 @@ class DictionarySegment : public BaseDictionarySegment {
 
  protected:
   const std::shared_ptr<const pmr_vector<T>> _dictionary;
-  const std::shared_ptr<const BaseCompressedVector> _attribute_vector;
+  std::shared_ptr<const std::span<const T>> _dictionary_span;
+  std::shared_ptr<const BaseCompressedVector> _attribute_vector;
```
Comment (klauck) [inline src/lib/storage/dictionary_segment.hpp:99]: Eventually the span should be named _dictionary. The corresponding data member maybe "dictionary_data"?

File: src/lib/storage/fixed_string_dictionary_segment.cpp [hunk 15]
```diff
@@ -37,17 +38,22 @@ std::optional<T> FixedStringDictionarySegment<T>::get_typed_value(const ChunkOff
...
-  return _dictionary->get_string_at(value_id);
+  return _dictionary_span->get_string_at(value_id);
 }
 
 template <typename T>
 std::shared_ptr<const FixedStringVector> FixedStringDictionarySegment<T>::fixed_string_dictionary() const {
   return _dictionary;
 }
 
+template <typename T>
```
Comment (klauck) [inline src/lib/storage/fixed_string_dictionary_segment.cpp:52]: Maybe, we can delete the old method `fixed_string_dictionary`.

File: src/lib/storage/dictionary_segment/dictionary_segment_iterable.hpp (hunk 13 above)
Comment (tjjordan) [reply src/lib/storage/dictionary_segment/dictionary_segment_iterable.hpp:15]: Of course.
But first, the memory of all segments need to be managed by the storage manager.

File: src/lib/storage/storage_manager.hpp (hunk 10 above)
Comment (tjjordan) [reply src/lib/storage/storage_manager.hpp:114]: This vector was not well thought through.
I just purged it.

File: src/lib/storage/dictionary_segment.hpp (hunk 14 above)
Comment (tjjordan) [reply src/lib/storage/dictionary_segment.hpp:99]: Good idea!

File: src/lib/storage/fixed_string_dictionary_segment.cpp (hunk 15 above)
Comment (tjjordan) [reply src/lib/storage/fixed_string_dictionary_segment.cpp:52]: Currently it's used inside the Hyrise Codebase.
We will remove it when the dictionary vectors get removed, but currently we don't want to touch too much parts of Hyrise at once.

Comment (Bouncner) [reply src/lib/storage/fixed_string_dictionary_segment.cpp:52]: How and why should we get rid of it?

There are operators that access the dictionary prior to the segment (e.g., for early exists). Do you mean removing accesses to segment dictionaries at all?

File: src/lib/storage/vector_compression/bitpacking/bitpacking_vector.cpp (hunk 2 above)
Comment (phoeinx) [reply src/lib/storage/vector_compression/bitpacking/bitpacking_vector.cpp]: This was a left-over from first try outs of how to access typed information of underlying attribute vectors.

**Pull request merged**